File: Turbine/EntityBuilder.fs
namespace Turbine

open System
open System.Collections.Generic
open Amazon.DynamoDBv2.Model

module internal EntityBuilder =

    let private hydrateFromProps
        (
            t: Type,
            instance: obj,
            attributes: IReadOnlyDictionary<string, AttributeValue>,
            schema: Schema
        ) =

        let partitionKeyProp =
            match schema.PartitionMap.TryGetValue t with
            | true, v -> v
            | _, _ -> schema.Pk

        let sortKeyProp =
            match schema.SortMap.TryGetValue t with
            | true, v -> v
            | _, _ -> schema.Sk

        let props = t.GetProperties()

        for prop in props do
            let propName = prop.Name
            let propType = prop.PropertyType

            let attributeValue =
                attributes
                |> Seq.tryFind (fun kvp -> kvp.Key.Equals(propName, StringComparison.OrdinalIgnoreCase))

            match attributeValue with
            | Some(KeyValue(_, av)) ->
                let value = Reflection.toNetType propType av
                prop.SetValue(instance, value)
            | None ->
                if propName.Equals(partitionKeyProp, StringComparison.OrdinalIgnoreCase) then
                    attributes
                    |> Seq.tryFind (fun kvp -> kvp.Key.Equals(schema.Pk, StringComparison.OrdinalIgnoreCase))
                    |> Option.map (fun (KeyValue(_, av)) ->
                        let value = Reflection.toNetType propType av
                        prop.SetValue(instance, value))
                    |> ignore

                if propName.Equals(sortKeyProp, StringComparison.OrdinalIgnoreCase) then
                    attributes
                    |> Seq.tryFind (fun kvp -> kvp.Key.Equals(schema.Sk, StringComparison.OrdinalIgnoreCase))
                    |> Option.map (fun (KeyValue(_, av)) ->
                        let value = Reflection.toNetType propType av
                        prop.SetValue(instance, value))
                    |> ignore

        instance

    let hydrateEntity<'T> (schema: Schema, attributes: IReadOnlyDictionary<string, AttributeValue>) : 'T =
        let entityType = typeof<'T>
        let ctors = entityType.GetConstructors()

        if ctors |> Seq.exists (fun c -> c.GetParameters().Length = 0) then
            let instance = Activator.CreateInstance(entityType)
            hydrateFromProps (entityType, instance, attributes, schema) |> unbox<'T>
        else
            failwith "not implemented"

File: Turbine/EntityBuilder.cs
using System.Reflection;
using Amazon.DynamoDBv2.Model;

namespace Turbine;

internal static class EntityBuilder
{
    private static object HydrateFromProps(
        Type t,
        object instance,
        IReadOnlyDictionary<string, AttributeValue> attributes,
        EntitySchema entitySchema)
    {
        var partitionKeyProp = entitySchema.GetPkProperty();
        var sortKeyProp = entitySchema.GetSkProperty();

        var props = Array.FindAll(t.GetProperties(), p => p.GetSetMethod(true) != null);

        foreach (var prop in props)
        {
            var matchingAttribute = attributes.KeyValueOrDefault(prop.Name);

            if (matchingAttribute is null && entitySchema.IsPkProperty(prop.Name))
            {
                matchingAttribute = attributes.KeyValueOrDefault(entitySchema.TableSchema.Pk);
            }

            if (matchingAttribute is null && entitySchema.IsSkProperty(prop.Name))
            {
                matchingAttribute = attributes.KeyValueOrDefault(entitySchema.TableSchema.Sk);
            }

            if (matchingAttribute is not null)
            {
                prop.SetValue(instance, Reflection.ToNetType(prop.PropertyType, matchingAttribute.Value.Value));
            }
        }

        return instance;
    }

    public static T HydrateEntity<T>(
        EntitySchema entitySchema,
        IReadOnlyDictionary<string, AttributeValue> attributes)
    {
        var entityType = typeof(T);
        var constructors = entityType.GetConstructors();

        if (constructors.Any(c => c.GetParameters().Length == 0))
        {
            return (T)HydrateFromProps(
                entityType,
                Activator.CreateInstance(entityType)!,
                attributes,
                entitySchema);
        }

        var instanceOpt =
            constructors
                .Select(c => (c, c.GetParameters()))
                .OrderByDescending(x => x.Item2.Length)
                .Select<(ConstructorInfo, ParameterInfo[]), (T?, Exception?)>(x =>
                {
                    var ctor = x.Item1;
                    var parameters = x.Item2;

                    var args = parameters
                        .Select(p =>
                        {
                            if (p.Name is null)
                            {
                                return null;
                            }

                            var matchingAttribute = attributes.KeyValueOrDefault(p.Name);

                            if (matchingAttribute is null && entitySchema.IsPkProperty(p.Name))
                            {
                                matchingAttribute = attributes.KeyValueOrDefault(entitySchema.TableSchema.Pk);
                            }

                            if (matchingAttribute is null && entitySchema.IsSkProperty(p.Name))
                            {
                                matchingAttribute = attributes.KeyValueOrDefault(entitySchema.TableSchema.Sk);
                            }

                            if (matchingAttribute is not null)
                            {
                                return Reflection.ToNetType(p.ParameterType, matchingAttribute.Value.Value);
                            }

                            return p.ParameterType.IsValueType ? Activator.CreateInstance(p.ParameterType) : null;
                        })
                        .ToArray();

                    try
                    {
                        return ((T)ctor.Invoke(args), null);
                    }
                    catch (Exception e)
                    {
                        return (default, e);
                    }
                })
                .FirstOrDefault();

        var (instance, exception) = instanceOpt;

        if (instance is not null)
        {
            return instance;
        }

        throw new TurbineException($"Could not create instance of '{entityType.Name}'.", exception);
    }
}